### PR TITLE
Add Spot/On-demand price to table-wide output irrespective of the price filters

### PR DESCRIPTION
**Pull request by mdb**
This seeks to address issue #71 . It features the original changes submitted by @krishna-birla in the unmerged (and perhaps abandoned?) PR #78 , but also layers in the outstanding changes requested by @bwagner5 in code review.

Note, however, that it doesn't yet tackle [this potential issue](https://github.com/aws/amazon-ec2-instance-selector/pull/78#issuecomment-829955937) pointed out by @krishna-birla, nor does it dramatically change @krishna-birla's original implementation.

~...this is currently a draft PR to surface visibility among the `amazon-ec2-instance-selector` maintainers and perhaps get some early feedback on what remaining work is necessary to get this to a mergeable state.~

Thanks!

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/ec2pricing/ec2pricing.go [hunk 1]
```diff
@@ -154,25 +184,27 @@ func (p *EC2Pricing) GetOndemandInstanceTypeCost(instanceType string) (float64,
...
-	// Check cache first and return it if available
-	if price, ok := p.cache[instanceType]; ok {
-		return price, nil
-	}
-
 	pricePerUnitInUSD := float64(-1)
-	err := p.PricingClient.GetProductsPages(&productInput, func(pricingOutput *pricing.GetProductsOutput, nextPage bool) bool {
-		var err error
+	var processingErr error
+	// FIXME Only works for us-east-1 and ap-south-1
```
Comment (bwagner5) [inline pkg/ec2pricing/ec2pricing.go:189]: this is fixed above in the constructor. The Pricing client is always initialized to us-east-1.

File: pkg/ec2pricing/ec2pricing.go [hunk 2]
```diff
@@ -154,25 +184,27 @@ func (p *EC2Pricing) GetOndemandInstanceTypeCost(instanceType string) (float64,
...
+	// https://docs.aws.amazon.com/awsaccountbilling/latest/aboutv2/using-pelong.html#pe-endpoint
+	errAPI := p.PricingClient.GetProductsPages(&productInput, func(pricingOutput *pricing.GetProductsOutput, nextPage bool) bool {
+		var errParse error
 		for _, priceDoc := range pricingOutput.PriceList {
-			_, pricePerUnitInUSD, err = parseOndemandUnitPrice(priceDoc)
+			_, pricePerUnitInUSD, errParse = parseOndemandUnitPrice(priceDoc)
 		}
-		if err != nil {
+		if errParse != nil {
```
Comment (bwagner5) [inline pkg/ec2pricing/ec2pricing.go:196]: I realize you didn't change this part... but it seems like this `if` should go in the for loop to catch all parse errors.

File: pkg/ec2pricing/ec2pricing.go [hunk 3]
```diff
@@ -234,17 +270,27 @@ func (p *EC2Pricing) HydrateOndemandCache() error {
 			{Type: aws.String(pricing.FilterTypeTermMatch), Field: aws.String("tenancy"), Value: aws.String("shared")},
 		},
 	}
-	err := p.PricingClient.GetProductsPages(&productInput, func(pricingOutput *pricing.GetProductsOutput, nextPage bool) bool {
+	var processingErr error
+	// FIXME Only works for us-east-1 and ap-south-1
```
Comment (bwagner5) [inline pkg/ec2pricing/ec2pricing.go:274]: should be good now.

**Review by bwagner5 — COMMENTED** [review]
This is looking nice! Thanks for picking this up!  I haven't checked it out and taken it for a test drive yet, but the code looks good.

**Comment by bwagner5** [comment]
Looks like `go mod tidy` needs to be run.

File: pkg/ec2pricing/ec2pricing.go (hunk 3 above)
Comment (mdb) [reply pkg/ec2pricing/ec2pricing.go]: Good call!

File: pkg/ec2pricing/ec2pricing.go (hunk 2 above)
Comment (mdb) [reply pkg/ec2pricing/ec2pricing.go]: Good catch!

**Comment by bwagner5** [comment]
looks like some unit-tests are failing having to do with pricing:

```
=== RUN   TestFilter_PricePerHour
selector_test.go:662: Should return 1 instance type; got 0

--- FAIL: TestFilter_PricePerHour (0.00s)
=== RUN   TestFilter_PricePerHour_NoResults
--- PASS: TestFilter_PricePerHour_NoResults (0.00s)
=== RUN   TestFilter_PricePerHour_OD
selector_test.go:701: Should return 1 instance type; got 0

--- FAIL: TestFilter_PricePerHour_OD (0.00s)
=== RUN   TestFilter_PricePerHour_Spot
selector_test.go:721: Should return 1 instance type; got 0
```

File: pkg/selector/selector_test.go [hunk 4]
```diff
@@ -657,10 +687,12 @@ func TestFilter_PricePerHour_NoResults(t *testing.T) {
 
 func TestFilter_PricePerHour_OD(t *testing.T) {
 	ec2Mock := setupMock(t, describeInstanceTypesPages, "t3_micro.json")
+	now := time.Now()
 	itf := selector.Selector{
 		EC2: ec2Mock,
 		EC2Pricing: &ec2PricingMock{
 			GetOndemandInstanceTypeCostResp: 0.0104,
+			lastOnDemandCacheUTC:            &now,
```
Comment (mdb) [inline pkg/selector/selector_test.go:695]: By ensuring the `*CacheUTC` field(s) are populated with a non-`nil` value, the tests exercise relevant code paths in `selector.rawFilter`.

File: pkg/selector/selector_test.go [hunk 5]
```diff
@@ -617,12 +635,22 @@ func (p *ec2PricingMock) HydrateSpotCache(days int) error {
...
+}
+
 func TestFilter_PricePerHour(t *testing.T) {
 	ec2Mock := setupMock(t, describeInstanceTypesPages, "t3_micro.json")
+	now := time.Now()
 	itf := selector.Selector{
 		EC2: ec2Mock,
 		EC2Pricing: &ec2PricingMock{
 			GetOndemandInstanceTypeCostResp: 0.0104,
+			lastOnDemandCacheUTC:            &now,
```
Comment (mdb) [inline pkg/selector/selector_test.go:653]: ☝️ By ensuring the `*CacheUTC` field(s) are populated with a non-`nil` value, the tests exercise relevant code paths in `selector.rawFilter`.

**Comment by bwagner5** [comment]
It looks like something isn't working quite right. I pulled the code down today and did a table-wide run and no prices were output. I haven't had a chance to dive any deeper. 

```
./ec2-instance-selector -o table-wide
Instance Type  VCPUs   Mem (GiB)  Hypervisor  Current Gen  Hibernation Support  CPU Arch  Network Performance  ENIs    GPUs    GPU Mem (GiB)  GPU Info  On-Demand Price/Hr  Spot Price/Hr (30d avg)
-------------  -----   ---------  ----------  -----------  -------------------  --------  -------------------  ----    ----    -------------  --------  ------------------  -----------------------
a1.2xlarge     8       16         nitro       false        false                arm64     Up to 10 Gigabit     4       0       0                        -Not Fetched-       -Not Fetched-
a1.4xlarge     16      32         nitro       false        false                arm64     Up to 10 Gigabit     8       0       0                        -Not Fetched-       -Not Fetched-
a1.large       2       4          nitro       false        false                arm64     Up to 10 Gigabit     3       0       0                        -Not Fetched-       -Not Fetched-
a1.medium      1       2          nitro       false        false                arm64     Up to 10 Gigabit     2       0       0                        -Not Fetched-       -Not Fetched-
a1.metal       16      32         none        false        false                arm64     Up to 10 Gigabit     8       0       0                        -Not Fetched-       -Not Fetched-
a1.xlarge      4       8          nitro       false        false                arm64     Up to 10 Gigabit     4       0       0                        -Not Fetched-       -Not Fetched-
c4.2xlarge     8       15         xen         true         true                 x86_64    High                 4       0       0                        -Not Fetched-       -Not Fetched-
c4.4xlarge     16      30         xen         true         true                 x86_64    High                 8       0       0                        -Not Fetched-       -Not Fetched-
c4.8xlarge     36      60         xen         true         true                 x86_64    10 Gigabit           8       0       0                        -Not Fetched-       -Not Fetched-
c4.large       2       3.75       xen         true         true                 x86_64    Moderate             3       0       0                        -Not Fetched-       -Not Fetched-
c4.xlarge      4       7.5        xen         true         true                 x86_64    High                 4       0       0                        -Not Fetched-       -Not Fetched-
c5.12xlarge    48      96         nitro       true         true                 x86_64    12 Gigabit           8       0       0                        -Not Fetched-       -Not Fetched-
c5.18xlarge    72      144        nitro       true         true                 x86_64    25 Gigabit           15      0       0                        -Not Fetched-       -Not Fetched-
c5.24xlarge    96      192        nitro       true         false                x86_64    25 Gigabit           15      0       0                        -Not Fetched-       -Not Fetched-
c5.2xlarge     8       16         nitro       true         true                 x86_64    Up to 10 Gigabit     4       0       0                        -Not Fetched-       -Not Fetched-
c5.4xlarge     16      32         nitro       true         true                 x86_64    Up to 10 Gigabit     8       0       0                        -Not Fetched-       -Not Fetched-
c5.9xlarge     36      72         nitro       true         true                 x86_64    10 Gigabit           8       0       0                        -Not Fetched-       -Not Fetched-
c5.large       2       4          nitro       true         true                 x86_64    Up to 10 Gigabit     3       0       0                        -Not Fetched-       -Not Fetched-
c5.metal       96      192        none        true         false                x86_64    25 Gigabit           15      0       0                        -Not Fetched-       -Not Fetched-
c5.xlarge      4       8          nitro       true         true                 x86_64    Up to 10 Gigabit     4       0       0                        -Not Fetched-       -Not Fetched-
NOTE: 362 entries were truncated, increase --max-results to see more
```

**Comment by mdb** [comment]
> It looks like something isn't working quite right.

@bwagner5 Thanks for looking. It appears to work for me using a fresh binary compiled via `make compile` -- is it possible one of us is confused?

```
$ ./build/ec2-instance-selector -o table-wide --region us-east-1

Instance Type  VCPUs   Mem (GiB)  Hypervisor  Current Gen  Hibernation Support  CPU Arch      Network Performance  ENIs    GPUs    GPU Mem (GiB)  GPU Info  On-Demand Price/Hr  Spot Price/Hr (30d avg)
-------------  -----   ---------  ----------  -----------  -------------------  --------      -------------------  ----    ----    -------------  --------  ------------------  -----------------------
a1.2xlarge     8       16         nitro       false        false                arm64         Up to 10 Gigabit     4       0       0                        $0.204              $0.0671
a1.4xlarge     16      32         nitro       false        false                arm64         Up to 10 Gigabit     8       0       0                        $0.408              $0.1343
a1.large       2       4          nitro       false        false                arm64         Up to 10 Gigabit     3       0       0                        $0.051              $0.02047
a1.medium      1       2          nitro       false        false                arm64         Up to 10 Gigabit     2       0       0                        $0.0255             $0.00841
a1.metal       16      32         none        false        false                arm64         Up to 10 Gigabit     8       0       0                        $0.408              $0.1343
a1.xlarge      4       8          nitro       false        false                arm64         Up to 10 Gigabit     4       0       0                        $0.102              $0.03386
c1.medium      2       1.69922    xen         false        false                i386, x86_64  Moderate             2       0       0                        $0.13               $0.013
c1.xlarge      8       7          xen         false        false                x86_64        High                 4       0       0                        $0.52               $0.23213
c3.2xlarge     8       15         xen         false        true                 x86_64        High                 4       0       0                        $0.42               $0.12853
c3.4xlarge     16      30         xen         false        true                 x86_64        High                 8       0       0                        $0.84               $0.28717
c3.8xlarge     32      60         xen         false        true                 x86_64        10 Gigabit           8       0       0                        $1.68               $0.54466
c3.large       2       3.75       xen         false        true                 i386, x86_64  Moderate             3       0       0                        $0.105              $0.0294
c3.xlarge      4       7.5        xen         false        true                 x86_64        Moderate             4       0       0                        $0.21               $0.05902
c4.2xlarge     8       15         xen         true         true                 x86_64        High                 4       0       0                        $0.398              $0.20069
c4.4xlarge     16      30         xen         true         true                 x86_64        High                 8       0       0                        $0.796              $0.27595
c4.8xlarge     36      60         xen         true         true                 x86_64        10 Gigabit           8       0       0                        $1.591              $0.66631
c4.large       2       3.75       xen         true         true                 x86_64        Moderate             3       0       0                        $0.1                $0.03453
c4.xlarge      4       7.5        xen         true         true                 x86_64        High                 4       0       0                        $0.199              $0.07209
c5.12xlarge    48      96         nitro       true         true                 x86_64        12 Gigabit           8       0       0                        $2.04               $0.82979
c5.18xlarge    72      144        nitro       true         true                 x86_64        25 Gigabit           15      0       0                        $3.06               $1.29916
```

**Comment by bwagner5** [comment]
You're right, I was definitely the one confused :)  I guess I was testing an old binary before... anyways LGTM! Thanks!

**Review by bwagner5 — APPROVED** [review]
/lgtm

**Pull request merged**